### DISCUSSION TITLE
How to store config info in kong

### DISCUSSION THREAD

**Discussion by natesh245** [Kong/kong#13524] [Help]
We have a requirement where we want to support multiple providers for use cases like sms / email. We want to store this provider info in a config which is accessible from kong and from  an admin app which can dynamically update the config.

We want to support multiple provider apis, fallback to a different api on issues and aggragate responses from multiple apis. All this has to happen based on the configurations defined. This config should be accessible from kong and a admin app which can modify this config.

Please suggest a efficient way to store and manage this config data in kong

**Comment by natesh245** [comment]
Any updates on this ?

**Comment by nowNick** [comment]
I'm afraid that in order to achieve this you're going to have to implement a custom plugin. Out of curiosity - how do you plan to aggregate the responses? I assume they will have different bodies, status codes, response headers etc. 🤔

**Reply by natesh245** [reply]
Haven't yet figured out the aggregation part yet. Will try to write a custom plugin as you suggested

**Comment by natesh245** [comment]
@nowNick thanks for the response

**Comment by natesh245** [comment]
@nowNick how can I store this configuration ? Can I store it as custom entities in datastore? This config should be accessible from kong as well as a admin app which can modify the config.

Or is there a more efficient way to store this config?

**Reply by nowNick** [reply]
Hmmm... that's a good question! I'd say using Kong as storage is not it's expected use case but maybe it can work for what you need. Definitely have a look at https://docs.konghq.com/gateway/latest/plugin-development/custom-entities/ 

I'm sorry but I'm missing some details on how your Admin API plans to interact with this configuration. Will it be a JSON that can be updated and then send as this custom plugin configuration? That again opens up security questions - is the Kong Admin API protected in your network? It all also depends on how Kong is deployed (that approach would work only on traditional / hybrid mode).

**Reply by natesh245** [reply]
By admin app I'm not referring to the kong admin api, It will be one  of our backend service

**Comment by lhanjian** [comment]
Why not use multiple `upstreams` ?
